cartography/intel/gcp/compute.py

```diff
@@ -0,0 +1,118 @@
+from cartography.intel.gcp.crm import get_gcp_projects
+
+logger = logging.getLogger(__name__)
+
+
+def get_zones_in_project(project_id, compute, max_results=None):
+    """
+    Return the zones where the Compute Engine API is enabled for the given project_id.
+    If the API is not enabled, return None.
+    :param project_id: The project id
+    :param compute: The compute resource object created by googleapiclient.discovery.build()
+    :param max_results: Optional cap on number of results returned by this function. Default = None, which means no cap.
+    :return: List of a project's zone objects if Compute API is turned on, else None.
+    """
+    try:
+        req = compute.zones().list(project=project_id, maxResults=max_results)
+        res = req.execute()
+        return res['items']
+    except HttpError as e:
+        reason = json.loads(e.content)['error']['errors'][0]['reason']
+        if reason == 'accessNotConfigured':
+            logger.debug(
+                (
+                    "Google Compute Engine API access is not configured for project %s. "
+                    "Full details: %s"
+                ),
+                project_id,
+                e
+            )
+            return None
+        elif reason == 'notFound':
+            logger.debug(
+                (
+                    "Project %s returned a 404 not found error. "
+                    "Full details: %s"
+                ),
+                project_id,
+                e
+            )
+            return None
+        else:
+            logger.error("Could not use Compute Engine API on project %s; Reason: %s".format(project_id, reason))
+            raise e
+
+
+def get_gcp_instances_in_project(project_id, compute):
+    """
+    Return list of all GCP instances in a given project regardless of zone.
+    :param project_id: The project id
+    :param compute: The compute resource object created by googleapiclient.discovery.build()
+    :return: List of all GCP instances in given project regardless of zone.
+    """
+    zones = get_zones_in_project(project_id, compute)
+    if not zones:
+        # If the Compute Engine API is not enabled for a project, there are no zones and therefore no instances.
+        return []
+    instances = []
+    for zone in zones:
+        req = compute.instances().list(project=project_id, zone=zone['name'])
+        res = req.execute()
+        zone_instances = res.get('items', [])
+        for instance in zone_instances:
+            # Include project_id and zone_name on the `instance` dict so we can load it more easily
+            instance['project_id'] = project_id
+            instance['zone_name'] = zone['name']
+            instances.append(instance)
+    return instances
+
+
+def load_gcp_instances(neo4j_session, data, gcp_update_tag):
+    query = """
+    MATCH (p:GCPProject{id:{ProjectId}})
+    MERGE (i:GCPInstance{id:{InstanceId}})
+    ON CREATE SET i.firstseen = timestamp()
```

Comment:
should we add the instanceid attribute? It's a dupe of the id but it's clearer and avoid users to know what we base our index/id on.

Reply:
probably do the same with GCPProject